src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -257,83 +201,106 @@ public bool ParseHeaders(TRequestHandler handler, ref SequenceReader<byte> reade
-                    if (readAhead > 0)
-                    {
-                        reader.Rewind(readAhead);
-                    }
+                    // Load next header line to parse as a span
+                    span = ParseMultiSpanHeader(ref reader);
 
-                    length = ParseMultiSpanHeader(handler, ref reader);
-                    if (length < 0)
+                    // If there any remaining line?
+                    if (length == -1 && span.Length == 0)
                     {
-                        // Not there
                         return false;
                     }
-
-                    reader.Advance(length);
-                    // As we crossed spans set the current span to default
-                    // so we move to the next span on the next iteration
-                    span = default;
                 }
             }
 
             return false;
         }
 
-        private int ParseMultiSpanHeader(TRequestHandler handler, ref SequenceReader<byte> reader)
+        // Returns the length of the header key:value and EOL.
+        // If no valid EOL is detected length = -1
+        // If the line is empty (only EOL) then length = 0
+        private int ParseHeaderLineEnd(ReadOnlySpan<byte> headerSpan, out int endLength)
+        {
+            endLength = -1;
+
+            var index = headerSpan.IndexOfAny(ByteCR, ByteLF);
+
+            if (index == -1)
+            {
+                return -1;
+            }
+
+            if (headerSpan[index] == ByteCR)
+            {
+                // No more chars after CR? Don't consume an incomplete header
+                if (headerSpan.Length == index + 1)
+                {
+                    return -1;
+                }
+
+                // CR must be followed by LF in all cases
+                if (headerSpan[index + 1] != ByteLF)
+                {
+                    if (index == 0)
+                    {
+                        KestrelBadHttpRequestException.Throw(RequestRejectionReason.InvalidRequestHeadersNoCRLF);
+                    }
+                    else
+                    {
+                        RejectRequestHeader(headerSpan[..(index + 2)]);
+                    }
+                }
+
+                endLength = 2;
+                return index;
+            }
+
+            if (_enableLineFeedTerminator)
+            {
+                endLength = 1;
+                return index;
+            }
+
+            // LF but not allowed
+            RejectRequestHeader(headerSpan[..(index + 1)]);
+
+            return -1;
+        }
+
+        // Returns a span from the remaining sequence until the next valid EOL
+        private ReadOnlySpan<byte> ParseMultiSpanHeader(ref SequenceReader<byte> reader)
         {
             var currentSlice = reader.UnreadSequence;
             var lineEndPosition = currentSlice.PositionOfAny(ByteCR, ByteLF);
```

Comment:
I think returning the `ReadOnlySpan<byte>` from `ParseMultiSpanHeader` is really smart. The one downside I see is that in the multi-span header case, we're now calling both `​PositionOfAny`​ and `​IndexOfAny`​ when we were just calling ​`PositionOfAny`​ in that case before. Is there a clean way to avoid this? Maybe a `out bool foundByteCR` or something?

Reply:
You don't even need that. I think you could just track if the `ReadOnlySpan<byte>` in `ParseHeaders` originated from `ParseMultiSpanHeader` and skip `IndexOfAny` and set `var index = headerSpan.Length` in `ParseHeaderLineEnd`.

Reply:
Done, was not as easy as expected, still simple logic.